Twig/Extension/SonataAdminExtension.php

```diff
@@ -326,22 +345,162 @@ public function getUrlsafeIdentifier($model, AdminInterface $admin = null)
+            'admin'            => $root->getCode(),
+            'associationAdmin' => $associationAdmin->getCode(),
+            'formType'         => $fieldDescription->getType(),
+            'editMode'         => $sonataAdmin['edit'],
+            'label'            => $associationAdmin->trans($associationAdmin->getLabel(), array(), $associationAdmin->getTranslationDomain()),
+            'routes'           => array(
+                'shortObjectDescription' => $router->generate('sonata_admin_short_object_information', array(
+                    'objectId'       => '__OBJECT_ID__',
+                    'uniqid'         => $associationAdmin->getUniqid(),
+                    'code'           => $associationAdmin->getCode(),
+                    'linkParameters' => $fieldDescription->getOption('link_parameters'),
+                )),
+                'retrieveFormElement' => $router->generate('sonata_admin_retrieve_form_element', array(
+                    'code'      => $root->getCode(),
+                    'elementId' => $context['id'],
+                    'objectId'  => $root->id($root->getSubject()),
+                    'uniqid'    => $root->getUniqid(),
+                    'subclass'  => $admin->getActiveSubclassCode(),
+                )),
+                'appendFormElement' => $router->generate(
+                    'sonata_admin_append_form_element',
+                    array_merge(array(
+                        'code'      => $root->getCode(),
+                        'elementId' => $context['id'],
+                        'objectId'  => $root->id($root->getSubject()),
+                        'uniqid'    => $root->getUniqid(),
+                        'subclass'  => $request->query->get('subclass'),
+                    ), $fieldDescription->getOption('link_parameters', array()))
+                ),
+            ),
+        );
+
+        return json_encode($data, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
+    }
+
+    public function renderAutocompleteOptions($context)
+    {
+        /** @var AppVariable $app */
+        $app = $context['app'];
+        $request = $app->getRequest();
+        $sonataAdmin = $context['sonata_admin'];
+        /** @var Admin $admin */
+        $admin = $sonataAdmin['admin'];
+        $router = $admin->getRouteGenerator();
+
+        // set the widget's initial value
+        $values = array();
+        $labels = isset($context['value']['_labels']) ? $context['value']['_labels'] : array();
+        unset($context['value']['_labels']);
+
+        foreach ($context['value'] as $i => $id) {
+            $values[] = array(
+                'id'    => $id,
+                'label' => $labels[$i],
+            );
+        }
+        // convert string values to booleans
+        $multiple = $context['multiple'] ? true : false;
+        $required = $context['required'] ? true : false;
+
+        $allowClearPlaceholder = (!$multiple && !$required) ? ' ' : '';
+
+        $data = array(
```

Comment:
The method is named `renderAutocompleteOptions`, that's why I was suggesting that : it should return auto-complete options. Gotta find a name that fits this big array of data that are somewhat related anyway. `$autocompleteAttributes, maybe` ? Or `$autcompleteConfiguration` ? It should convey the following (correct me if I'm wrong) : 

- this is about configuring a widget
- the widget is an autocomplete widget